src/history/hs_rec.c

```diff
@@ -912,6 +912,48 @@ __hs_delete_reinsert_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, ui
         if (__wt_txn_tw_stop_visible_all(session, twp))
             continue;
 
+        /*
+         * The below condition is to handle scenario where the data store and the
+         * history store has same records and ignore.
+         *
+         * Suppose there are two tables table1 and table2 and the below operations are performed.
+         *
+         * 1. Insert a=1 in table1 at timestamp 10
+         * 2. Insert b=1 in table2 at timestamp 10
+         * 3. Delete a from table1 at timestamp 20
+         * 4. Delete b from table2 at timestamp 20
+         * 5. Set stable timestamp = 20, oldest timestamp=1
+         * 6. Checkpoint table1
+         * 7. Checkpoint table2
+         * 8. Insert a=2 in table1 at timestamp 30
+         * 9. Insert b=2 in table2 at timestamp 30
+         * 10. Evict a=2 from table1 and move the content to history store.
+         * 11. Checkpoint history store.
+         *
+         * After all this operations the checkpoint content will be
+         * Data store --
+         * table1 --> a=1 at start_ts=10, stop_ts=20
+         * table2 --> b=1 at start_ts=10, stop_ts=20
+         *
+         * History store --
+         * table1 --> a=1 at start_ts=10, stop_ts=20
+         *
+         * WiredTiger takes a backup of the checkpoint and use this backup to restore.
+         * Note: In table1 of both data store and history store has the same content.
+         *
+         * Now the backup is used to restore.
+         *
+         * 1. Insert a=3 in table1
+         * 2. Checkpoint started, eviction started and sees the same content in data store and
+         * history while reconciling.
+         */
+
+        if (upd_tw != NULL &&
```

Comment:
We should also add the comment why we need to first check globally visibility of the start.

Reply:
I think it should be

__wt_txn_tw_start_visible_all(session, upd_tw) && __wt_txn_tw_start_visible_all(session, twp) ? .....
Otherwise, we may upd_tw is globally visible and the stop is the same but the start is actually different.

Reply:
Fixed